lib/webview-api.js

```diff
@@ -0,0 +1,194 @@
+          try {
+            query.pluginArgs = JSON.parse(query.pluginArgs)
+          } catch (err) {
+            query.pluginArgs = []
+          }
+          query.pluginArgs.unshift(query.pluginAction)
+          this.utils.emit.apply(this, query.pluginArgs)
+        }
+        this.utils.emitEvent('did-navigate-in-page', [
+          {},
+          webView
+            .windowScriptObject()
+            .evaluateWebScript('window.location.href')
+        ])
+      },
+
+      // // Called when the JavaScript window object in a frame is ready for loading.
+      // 'webView:didClearWindowObject:forFrame:': function (
+      //   webView,
+      //   windowObject,
+      //   webFrame
+      // ) {},
+
+      // // Called when content starts arriving for a page load.
+      // 'webView:didCommitLoadforFrame:': function (webView, webFrame) {},
+
+      // // Notifies the delegate that a new JavaScript context has been created.
+      // 'webView:didCreateJavaScriptContext:forFrame:': function (
+      //   webView,
+      //   context,
+      //   webFrame
+      // ) {
+
+      // },
+
+      // Called when an error occurs loading a committed data source.
+      'webView:didFailLoadWithError:forFrame:': function (
+        webView,
+        error,
+        webFrame
+      ) {
+        this.utils.emit('did-fail-load', error)
+      },
+
+      // Called when a page load completes.
+      'webView:didFinishLoadForFrame:': function (webView, webFrame) {
+        if (this.state.wasReady == 0) {
+          this.utils.emitBrowserEvent('ready-to-show')
+          this.state.setObject_forKey(1, 'wasReady')
+        }
+        this.utils.emit('did-finish-load')
+        this.utils.emit('did-frame-finish-load')
+        this.utils.emit('dom-ready')
+      },
+
+      // // Called when a provisional data source for a frame receives a server redirect.
+      // 'webView:didReceiveServerRedirectForProvisionalLoadForFrame:': function (
+      //   webView,
+      //   webFrame
+      // ) {
+      //   this.utils.emit('did-get-redirect-request')
+      // },
+
+      // // Called when the page title of a frame loads or changes.
+      // 'webView:didReceiveTitle:forFrame:': function (
+      //   webView,
+      //   title,
+      //   webFrame
+      // ) {
+      //   this.utils.emitBrowserEvent('page-title-updated', {
+      //     preventDefault () {
+      //       // TODO: can we actually prevent it?
+      //     }
+      //   }, title)
+      // },
+
+      // // Called when a page load is in progress in a given frame.
+      // 'webView:didStartProvisionalLoadForFrame:': function (
+      //   webView,
+      //   webFrame
+      // ) {
+      //   this.utils.emit('did-start-loading')
+      // },
+
+      // Called when a frame will be closed.
+      'webView:willCloseFrame:': function (webView, webFrame) {
+        this.utils.emit('will-prevent-unload', {
+          preventDefault: function () {
+            this.utils.close()
```

Comment:
I suspect this is either a bug, or I don't understand what it's trying to do. My interpretation is that I should be able to do this:

```JavaScript
this.ui.webContents.on('will-prevent-unload', (event) => {
  event.preventDefault()
  console.log('The window did not close')
})
```
However this has the opposite effect. In fact, it appears to be executed on launch, which is a bit confusing.

Reply:
hum indeed, I'm not sure what I wanted to do here 🤔 

based on this: https://github.com/electron/electron/blob/master/docs/api/web-contents.md#event-will-prevent-unload, this event is to be fired when the webview is prevented to be closed. I'm not sure how to do that

Reply:
So I think it's the `close` event that should have a `preventDefault` method which, when called, should trigger the `will-prevent-unload` with a `preventDefault` that still closes the webview

Reply:
Ah yes, that makes sense. Based on what you mentioned, this works:

```
this.ui.on('close', (event) => {
  this.ui.hide()
  event.preventDefault()
});
```

It appears (from my test) that nothing executes after `event.preventDefault()`. This feels odd since I'm used to putting it first in the handler, but, assuming I'm not just doing something wrong or there's a bug, I can live with it.

Reply:
> It appears (from my test) that nothing executes after `event.preventDefault()`

because `event` is undefined ;) when I said "should", I meant that it should do that but doesn't yet haha

Reply:
Oh, ha. I'm surprised I didn't notice it throw an error. That's something I ought to investigate myself. Thanks for resolving it.